lib/forgetsy/set.rb

```diff
@@ -58,90 +58,87 @@ def fetch(opts = {})
       scrub if opts.fetch(:scrub, true)
 
       if opts.key?(:bin)
-        result = [opts[:bin], @conn.zscore(@name, opts[:bin])]
+        result = [[opts[:bin], redis.zscore(@name, opts[:bin])]]
       else
         result = fetch_raw(limit: limit)
       end
 
-      Hash[*result.flatten]
+      Hash[result.map{ |r| [r[0], r[1]] }]
     end
 
     # Apply exponential time decay and
     # update the last decay time.
     def decay(opts = {})
+      last_decayed_date, lifetime = last_decayed_date_and_lifetime
       t0 = last_decayed_date
       t1 = opts.fetch(:date, Time.now).to_f
       delta_t = t1 - t0
       set = fetch_raw
       rate = 1 / Float(lifetime)
-      @conn.pipelined do
+      redis.pipelined do
         set.each do |k, v|
           new_v = v * Math.exp(- delta_t * rate)
-          @conn.zadd(@name, new_v, k)
+          redis.zadd(@name, new_v, k)
         end
         update_decay_date(Time.now)
       end
     end
 
     def scrub
-      @conn.zremrangebyscore(@name, '-inf', @@hi_pass_filter)
+      redis.zremrangebyscore(@name, "-inf".freeze, HIGH_PASS_FILTER)
     end
 
     def incr(bin, opts = {})
       date = opts.fetch(:date, Time.now)
-      @conn.zincrby(@name, 1, bin) if valid_incr_date(date)
+      redis.zincrby(@name, 1, bin) if valid_incr_date(date)
     end
 
     def incr_by(bin, by, opts = {})
       date = opts.fetch(:date, Time.now)
-      @conn.zincrby(@name, by, bin) if valid_incr_date(date)
+      redis.zincrby(@name, by, bin) if valid_incr_date(date)
     end
 
     def last_decayed_date
-      @conn.zscore(@name, @@last_decayed_key)
+      redis.hget(METADATA_KEY, metadata_key(LAST_DECAYED_KEY)).to_f
```

Comment:
While i'm not opposed to separating out the metadata into its own top-level key, this is a breaking change in that both `last_decayed_date` and `lifetime` will not report correctly in the event of an upgrade. Upgrading will require a data migration. I'm not sure it's worth introducing a breaking change for this.